crypto/hpke/hpke.c

```diff
@@ -0,0 +1,1537 @@
+{
+    if (ctx == NULL)
+        return 0;
+    if (ctx->authpub != NULL)
+        OPENSSL_free(ctx->authpub);
+    ctx->authpub = OPENSSL_malloc(publen);
+    if (ctx->authpub == NULL)
+        return 0;
+    memcpy(ctx->authpub, pub, publen);
+    ctx->authpublen = publen;
+    return 1;
+}
+
+int OSSL_HPKE_CTX_get_seq(OSSL_HPKE_CTX *ctx, uint64_t *seq)
+{
+    if (ctx == NULL || seq == NULL)
+        return 0;
+    *seq = ctx->seq;
+    return 1;
+}
+
+int OSSL_HPKE_CTX_set_seq(OSSL_HPKE_CTX *ctx, uint64_t seq)
+{
+    if (ctx == NULL)
+        return 0;
+    ctx->seq = seq;
+    return 1;
+}
+
+int OSSL_HPKE_encap(OSSL_HPKE_CTX *ctx,
+                    unsigned char *enc, size_t *enclen,
+                    const unsigned char *pub, size_t publen,
+                    const unsigned char *info, size_t infolen)
+{
+    int erv = 1;
+
+    if (ctx == NULL || enc == NULL || enclen == NULL || *enclen == 0
+        || pub == NULL || publen == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (ctx->shared_secret != NULL) {
+        /* only allow one encap per OSSL_HPKE_CTX */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    if (hpke_encap(ctx, enc, enclen, pub, publen) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /*
+     * note that the info is not part of the context as it
+     * only needs to be used once here so doesn't need to
+     * be stored
+     */
+    erv = hpke_do_middle(ctx, info, infolen);
+    return erv;
+}
+
+int OSSL_HPKE_decap(OSSL_HPKE_CTX *ctx,
+                    const unsigned char *enc, size_t enclen,
+                    EVP_PKEY *recippriv,
+                    const unsigned char *info, size_t infolen)
+{
+    int erv = 1;
+
+    if (ctx == NULL || enc == NULL || enclen == 0 || recippriv == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+    if (ctx->shared_secret != NULL) {
+        /* only allow one encap per OSSL_HPKE_CTX */
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+    erv = hpke_decap(ctx, enc, enclen, recippriv);
+    if (erv != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    /*
+     * note that the info is not part of the context as it
+     * only needs to be used once here so doesn't need to
+     * be stored
+     */
+    erv = hpke_do_middle(ctx, info, infolen);
+    return erv;
+}
+
+int OSSL_HPKE_seal(OSSL_HPKE_CTX *ctx,
+                   unsigned char *ct, size_t *ctlen,
+                   const unsigned char *aad, size_t aadlen,
+                   const unsigned char *pt, size_t ptlen)
+{
+    int erv = 1;
+    unsigned char seqbuf[12];
+    size_t seqlen = 0;
+    unsigned char nonce[12];
+
+    if (ctx == NULL || ct == NULL || ctlen == NULL || *ctlen == 0
```

Comment:
Maybe this should handle the case of ct=NULL and ctlen != NULL by returning the required size. (This is a common pattern in openssl).

Reply:
Maybe. Will come back to it as that'd change the API/POD file too and calling code.

Reply:
So, I guess that could be done, but I'd rather not tbh - there're already APIs that allow the caller to know output sizes and if we made the change suggested above we'd really have to do the same for the all 5 of functions with buffer outputs. (IIRC we discussed that topic a few months back with others and I didn't get the sense that this design pattern is a requirement.)